File: LeetCode/ProblemSet/Volume (3500-3599)/LeetCode3532/LeetCode3532-Solution-01.go
package leetcode3532

func pathExistenceQueries(n int, nums []int, maxDiff int, queries [][]int) []bool {
	pre := make([]int, n)
	pre[0] = 0
	for i := 1; i < n; i++ {
		if nums[i] - nums[i - 1] <= maxDiff {
			pre[i] = pre[i - 1] + 1
		} else {
			pre[i] = pre[i - 1]
		}
	}

	ans := make([]bool, len(queries))
	for i, query := range queries {
		u, v := query[0], query[1]
		ans[i] = pre[v] - pre[u] == v - u
	}

	return ans
}

File: LeetCode/ProblemSet/Volume (3500-3599)/LeetCode3532/LeetCode3532-Solution-01.cpp
#include <bits/stdc++.h>

class Solution {
public:
    std::vector<bool> pathExistenceQueries(int n, std::vector<int>& nums, int maxDiff, std::vector<std::vector<int>>& queries) {
        std::vector<int> pre(n);

        pre[0] = 0;
        for (int i = 1; i < n; i++) {
            pre[i] = pre[i - 1] + (std::abs(nums[i] - nums[i - 1]) <= maxDiff ? 1 : 0);
        }

        std::vector<bool> ans;
        for (auto &query : queries) {
            int u = query[0], v = query[1];

            if (u == v) {
                ans.emplace_back(true);
                continue;
            } 
           
            bool res = pre[v] - pre[u] == v - u;
            ans.emplace_back(res);
        }

        return ans;
    }
};